File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution(object):
    def isValid(self, s):
        stack = []
        for b in s:
            if b in ('(' , '[' , '{'):
                stack.append(b)
            elif len(stack):
               if abs(ord(stack.pop()) - ord(b)) not in (1,2):
                   return False                      
            else:
                return False

        if not len(stack):
            return True
        
        return False
        
# @lc code=end

File: 20.valid-parentheses.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */

// @lc code=start

impl Solution {
    pub fn is_valid(input_str: String) -> bool {
        let mut stack: Vec<char> = Vec::new();

        for curr_char in input_str.chars() {
            match curr_char {
                '(' | '{' | '[' => {
                    stack.push(curr_char);
                }

                ')' | '}' | ']' => match stack.pop() {
                    Some(popped_char) => match curr_char as i32 - popped_char as i32 {
                        2 | 1 => {
                            continue;
                        }
                        _ => {
                            return false;
                        }
                    },

                    None => {
                        return false;
                    }
                },

                _ => {}
            }
        }
        if stack.len() != 0 {
            return false;
        } else {
            true
        }
    }
}
// @lc code=end

